Addresses Issue #3138 - ESO - Enable Table Access Protocol (TAP) queries

juanmcloaiza opened the pull request:
Closes #3138

### Changes:

Switch querying interface from WDB to TAP in querying functions:
- [x] query_main
- [x] query_instruments
- [x] query_surveys
- [x] query_apex_quicklooks
- [x] list_instruments
- [x] list_surveys

Deprecate arguments:
- [x] `open_form`
- [x] `cache`

New functionality:
- [x] Allow plain ADQL queries via `query_tap_service` (with authentication as well) 
- [x] Cone search using `cone_ra`, `cone_dec`, `cone_radius` arguments
- [x] Retrieve record count before querying the archive, via `count_only` argument
- [x] Ask query functions to print the underlying ADQL queries without issuing them

juanmcloaiza commented:
Hi @bsipocz , could you please give us a time estimate of the review process? No pressure at all though, we only need to plan our activities internally. Thanks a lot in advance.

bsipocz commented on astroquery/eso/utils.py:
```diff
@@ -0,0 +1,197 @@
+        )
+
+
+def _py2adql(user_params: _UserParams) -> str:
```
can we call this something more descriptive? `parameters_to_adql` or something similar?

juanmcloaiza replied:
Renamed to `_build_adql_string`

bsipocz commented on astroquery/eso/core.py:
```diff
@@ -296,248 +218,472 @@ def _login(self, *, username: str = None, store_password: bool = False,
+
+        query = _py2adql(up)
+
+        if up.query_str_only:
```
have a `get_query_payload` argument instead of this to be consistent with the rest of the API

bsipocz commented on astroquery/eso/utils.py:
```diff
@@ -0,0 +1,197 @@
+    return column_list
+
+
+def _raise_if_has_deprecated_keys(filters: Optional[Dict[str, str]]) -> bool:
```
No need for this, the deprecation warnings should be raised the standard way

juanmcloaiza replied:
Not really... this is part of a larger problem that makes me uneasy: **The underlying querying language (wdb, tap, or whatever) is creeping up to the python API**... I'd like to address this problem in an upcoming pull request. I think that the API should be strictly python, no semi-strings related to the underlying querying tool... i.e., either users go all-in writing an ADQL query via `query_tap(...)` or they use the python API with the available 100% python arguments... 

Now, focusing on the purpose of this function in the current API: the user can pass a very valid dictionary to the query functions, for example `{... , 'box' = 'whatever', ...}`. However, the keyword `box` is a WDB keyword not available in TAP. As far as the **python** API is concerned, there's no deprecation since the dictionary is well formed, but TAP will surely complain. This special function is a courtesy to our users, to tell them what went wrong and how to fix. If you insist, we can remove the function... is not strictly necessary and I wouldn't lose our time discussing it further.

bsipocz commented on astroquery/eso/core.py:
```diff
@@ -296,248 +218,472 @@ def _login(self, *, username: str = None, store_password: bool = False,
+        astropy.conf.max_lines = n_
+        astropy.conf.max_width = m_
+
+    def _query_on_allowed_values(
```
I'm not convinced that this method is necessary, `list_columns` should deal with the help, the deprecations are handled by the deprecation decorator, the coordinates can be ducktyped (or run through a SkyCoord construction) and then the query can be called in the method that calls this one.

juanmcloaiza replied:
You're right, and indeed, an upcoming pull request is planed to greatly improve the way the code is written. This PR focuses on switching the backend from WDB to TAP trying to impact as little as possible the API and the code structure (and it still grew larger than expected). Please be patient with these kind of apparently un-needed methods for the moment. They will disappear very soon.

bsipocz commented on astroquery/eso/core.py:
```diff
@@ -296,248 +218,472 @@ def _login(self, *, username: str = None, store_password: bool = False,
+
+    @deprecated_renamed_argument(('open_form', 'cache'), (None, None),
+                                 since=['0.4.11', '0.4.11'])
+    def query_main(
```
maybe this is a good time to rename this method to be more descriptive? Maybe `query_instruments`?

Then we can fully deprecate the single instrument query from below, as that would just be a special case for running this query

bsipocz commented on astroquery/eso/core.py:
```diff
@@ -296,248 +218,472 @@ def _login(self, *, username: str = None, store_password: bool = False,
+
+    @deprecated_renamed_argument(('open_form', 'cache'), (None, None),
+                                 since=['0.4.11', '0.4.11'])
+    def query_instrument(
```
is there a need for this now as it feels to be just a special case for `query_main` above

juanmcloaiza replied:
query_instruments and query_main query from two different data sources, so they are both needed.

bsipocz commented on astroquery/eso/core.py:
```diff
@@ -606,10 +752,10 @@ def get_headers(self, product_ids, *, cache=True):
                     column_types.append(type(header[key]))
         # Add all missing elements
-        for i in range(len(result)):
+        for r in result:
```
nitpicking, but please try to avoid single letter variables

juanmcloaiza replied:
renamed to `item`

bsipocz commented on astroquery/eso/core.py:
```diff
@@ -667,14 +813,13 @@ def _download_eso_files(self, file_ids: List[str], destination: Optional[str],
                 else:
                     log.error(f"Failed to download {file_link}. {http_error}")
-            except Exception as ex:
+            except RuntimeError as ex:
```
Thanks for fixing this! We should indeed avoid catching generic `Exceptions`

bsipocz commented on astroquery/eso/core.py:
```diff
@@ -769,9 +915,11 @@ def get_associated_files(self, datasets: List[str], *, mode: str = "raw",
-    @deprecated_renamed_argument(('request_all_objects', 'request_id'), (None, None), since=['0.4.7', '0.4.7'])
-    def retrieve_data(self, datasets, *, continuation=False, destination=None, with_calib=None,
-                      request_all_objects=False, unzip=True, request_id=None):
+    @deprecated_renamed_argument(('request_all_objects', 'request_id'), (None, None),
+                                 since=['0.4.7', '0.4.7'])
```
This is a bit unrelated so no need to do it in this PR, but 0.4.7 was a while ago, so these deprecations can be now removed.

bsipocz commented on astroquery/eso/utils.py:
```diff
@@ -0,0 +1,197 @@
+    return where_constraints
+
+
+def reorder_columns(table: Table,
```
I have multiple questions here:

- is this really necessary? Why the default column order is not suffucient?
- but if this is necessary, then please just go by suggested way for table column reordering from the astropy docs, "Select or reorder columns" on this page: https://docs.astropy.org/en/stable/table/modify_table.html

Also, minor nitpicking, but all the other functions here are private, why is this one not?

juanmcloaiza replied:
Hmmm... but the function is already doing exactly what the docs suggest (isn't it?)...

From the docs:
```
new_order = ['a', 'c', 'b']  # List or tuple
t_acb = t[new_order]
```

From the ESO module:
```
[...]
last_cols = first_cols + last_cols  # this is a list of column names
table = table[last_cols]
```

The code above those two lines is just computing the list of column names.

Other than that,
 - the function is now private, good point.
 - yes, the reordering is necessary, because the default is alphabetical order and the first fields are long urls and other info that is not legible and take much space. The reorder was suggested by several beta testers, which were expecting the columns in the order that this function generates.

bsipocz commented on astroquery/eso/utils.py:
```diff
@@ -0,0 +1,197 @@
+    ['object', 'ra', 'dec', 'dp_id', 'date_obs']
+    Returns a table with the columns reordered.
+    """
+    if not isinstance(table, Table):
+        return table
```
Is this an expected case?

juanmcloaiza replied:
yes, `table` may be a string or an int, depending on the user parameters `query_str_only` only and `count_only`.

I admit this is not the best software design at this point, but a complete refactoring would have taken more time from both developers and reviewers. We will come with better encapsulation, separation of concerns,  etc in the next pull request. Our aim for now was to switch to TAP with as little impact as possible to the API and to the overall code.

bsipocz commented on astroquery/eso/tests/test_eso_remote.py:
I haven't reviewed the tests, but running them returned 77 failures, so please have a look.

`pytest -P eso -R` should work if you have all the test dependencies installed.

bsipocz commented on docs/eso/eso.rst:
```diff
@@ -4,6 +4,33 @@
 ESO Queries (`astroquery.eso`)
 ******************************
 
+.. warning::
```
This is very good! Thanks for adding.

bsipocz reviewed (commented):
I'm sorry to have this many comments and thank you for your patience while waiting for this review.

I acknowledge that this seems to a lot, but most of them are minor and/or helping API consistency with the other modules and/or making the code cleaner and easier to read and follow; which will be helpful for the future maintenance.

juanmcloaiza commented on astroquery/eso/core.py:
```diff
@@ -1,59 +1,59 @@
-                                 "problem; the service may be offline.")
-    elif b"# No data returned !" not in content:
-        return True
+__doctest_skip__ = ['EsoClass.*']
```
I know this is not a valid answer, but "because it was so in the main branch, L30", so I left it like that. I'll leave this for later, when I have a valid answer.

juanmcloaiza commented on astroquery/eso/core.py:
```diff
@@ -62,168 +62,90 @@ def _get_exp_time_from_token(self) -> int:
 
-    def _activate_form(self, response, *, form_index=0, form_id=None, inputs={},
-                       cache=True, method=None):
+    @property
+    def maxrec(self):
```
We also discussed this point internally. While both `maxrec` and `top` limit the number of results, their intent is different. `top` is part of the ADQL query, used limit the result size from a logical/scientific standpoint. `maxrec`, on the other hand, is a TAP server parameter used as a safety cap regardless of the query’s intent. For example, I'd use `maxrec` to avoid system overload when I'm on my laptop, and for quickly testing the notebook’s logic and plots before the "real-thing-run" that could last some hours.

Since `maxrec` reflects user's environment rather than query logic, it makes more sense to have it as a configurable property, rather than a per-query argument (as `top`). Regarding other modules: if they don’t yet distinguish between `maxrec` and `top`, I’d warmly suggest they do, since it’s an important conceptual separation.

juanmcloaiza replied:
I'm noticing this just now. The reply is here: https://github.com/astropy/astroquery/pull/3339#issuecomment-3327400641, `maxrec` was substituted by `ROW_LIMIT`, and yes,

```
eso_max50 = EsoClass()
eso_max50.maxrec = 50
```

This works 👍

juanmcloaiza commented on astroquery/eso/core.py:
```diff
@@ -62,168 +62,90 @@ def _get_exp_time_from_token(self) -> int:
+
+        self._maxrec = mr
+
+    def _tap_url(self) -> str:
```
I see... this is getting the `tap_url` from the config *UNLESS* the env var `ESO_TAP_URL` is set. This is mostly for debugging purposes: to use a test server rather than the one in the configuration. This is transparent for the user, but if you have a strong position against it, I can just remove it:

Change

```
    def _tap_url(self) -> str:
        url = os.environ.get('ESO_TAP_URL', conf.tap_url)
        return url
```

for

```
_tap_url = conf.tap_url
```

And propagate that change throughout the module.... or use conf.tap_url directly all-around.

juanmcloaiza commented on astroquery/eso/core.py:
```diff
@@ -296,248 +218,472 @@ def _login(self, *, username: str = None, store_password: bool = False,
+        --------
+        >>> from astroquery.eso import Eso
+        >>> eso_instance = Eso()
+        >>> eso_instance.query_tap_service("SELECT * FROM ivoa.ObsCore")
+        """
+        table_to_return = None
+        tap_service = self._tap_service(authenticated)
+        table_to_return = self._try_download_pyvo_table(query_str, tap_service)
+        return table_to_return
```
This is relevant for some of your previous comments as well. I'm only marking it for the moment to take a closer look to all related comments together.

juanmcloaiza commented on astroquery/eso/core.py:
```diff
@@ -296,248 +218,472 @@ def _login(self, *, username: str = None, store_password: bool = False,
+        self,
+        user_params: _UserParams
+    ) -> Union[astropy.table.Table, int, str]:
+        up = user_params  # shorthand
```
Well, I can assure you it made the code more writeable 😄  ... but you're right -- shorthand removed.

juanmcloaiza commented on astroquery/eso/core.py:
```diff
@@ -296,248 +218,472 @@ def _login(self, *, username: str = None, store_password: bool = False,
+
+        return table_to_return
+
+    def _tap_service(self, authenticated: bool = False) -> TAPService:
```
For the moment, I am renaming the function `_tap_service` to `tap`. Making it a property would have repercussions in other parts of the code which at this point I'd avoid touching. I would leave these larger refactoring changes to a next Pull Request, where we plan a larger refactoring, if you agree.

juanmcloaiza commented on astroquery/eso/core.py:
```diff
@@ -296,248 +214,468 @@ def _login(self, *, username: str = None, store_password: bool = False,
+        # This whole function should be better written and the output
+        # shown tidier
+        nlines = len(available_cols) + 2
+        n_ = astropy.conf.max_lines
+        m_ = astropy.conf.max_width
+        astropy.conf.max_lines = nlines
+        astropy.conf.max_width = sys.maxsize
+        log.info(f"\nColumns present in the table {table_name}:\n{available_cols}\n"
+                 f"\nNumber of records present in the table {table_name}:\n{num_records}\n")
+        astropy.conf.max_lines = n_
+        astropy.conf.max_width = m_
```
Nice suggestion! I'll try it 👍

juanmcloaiza commented on astroquery/eso/core.py:
```diff
@@ -801,15 +939,18 @@ def retrieve_data(self, datasets, *, continuation=False, destination=None, with_
+        files = Eso.retrieve_data(dpids)
 
         """
+        _ = request_all_objects, request_id
```
This function is different than the rest... 

On the `main` branch, the signature looks like this:

```
    @deprecated_renamed_argument(('request_all_objects', 'request_id'), (None, None), since=['0.4.7', '0.4.7'])
    def retrieve_data(self, datasets, *, continuation=False, destination=None, with_calib=None,
                      request_all_objects=False, unzip=True, request_id=None):
        """
        Retrieve a list of datasets form the ESO archive.
...
```

Then the `request_all_objects` and `request_id` are ignored and not used anymore throughout the function. There's not even a hint to the developers that those arguments are being ignored (Like the underscores I use).

I think that, in this case, since the arguments were deprecated already in version 0.4.7, we can remove the deprecation decorator AND also remove the arguments from the signature altogether. (In fact, the deprecation decorator was already removed, [see here](https://github.com/astropy/astroquery/pull/3339#discussion_r2244201212)).

Unless you object, I'd remove the arguments altogether.

juanmcloaiza commented:
@keflavich, your comments have been taken into account, thanks for having taken the time. [cc @bsipocz, @ashleythomasbarnes, @szampier].

juanmcloaiza commented:
@keflavich, @bsipocz, just FYI, on making the requested changes I introduced a bug that is now fixed in commit [aef5c0f](https://github.com/astropy/astroquery/pull/3339/commits/aef5c0f2352adf9bd10aafa737872042a702b8d9) [@ashleythomasbarnes, @szampier].

bsipocz commented on astroquery/eso/core.py:
```diff
@@ -1,59 +1,55 @@
-        return True
+from .utils import _UserParams, raise_if_coords_not_valid, _reorder_columns, \
+    _raise_if_has_deprecated_keys, _build_adql_string, \
+    DEFAULT_LEAD_COLS_PHASE3, DEFAULT_LEAD_COLS_RAW
```
have an all to keep the tab complete clean and to signal the public API.
```suggestion
    DEFAULT_LEAD_COLS_PHASE3, DEFAULT_LEAD_COLS_RAW
    
    
__all__ = ['Eso', 'EsoClass']
```

bsipocz commented:
RTD failure is unrelated and is due to a recent plugin release which I'll resolved in a separate PR.
